File: E-EE-E/programmers/target_number.js
//https://programmers.co.kr/learn/courses/30/lessons/43165
let numbers = [1,1,1,1,1], target = 3;
console.log(solution(numbers, target));

function solution(numbers, target) {
    var answer = 0;
    let set = [];
    function dfs(){
        if(set.length==numbers.length){
            let result = 0;
            for(let i =0;i<numbers.length;i++) result+=numbers[i]*set[i];
            if(result==target) answer++;
            return;
        }
        set.push(1);
        dfs();
        set.pop();
        set.push(-1);
        dfs();
        set.pop();
        return;
    }
    dfs();
    return answer;
}

File: E-EE-E/programmers/target_number.cpp
//https://programmers.co.kr/learn/courses/30/lessons/43165?language=cpp
#include "iostream"
#include <string>
#include <vector>
#include <map>
using namespace std;

int solution(vector<int> numbers, int target) {
    map<int, int> arr, temp;
    arr[0]=1;
    for(auto x : numbers){
        temp.clear();
        for(auto y : arr){
            temp[y.first+x]+=y.second;
            temp[y.first-x]+=y.second;
        }
        arr=temp;
    }
    int answer = arr[target];
    return answer;
}

int main (){
    vector<int> numbers = { 1, 1, 1, 1, 1};
    int target = 3;
    cout<<solution(numbers, target);    // == 5
}